### DISCUSSION TITLE
WARNING: Deterministic matching rules led to no observed matches!

### DISCUSSION THREAD

**Discussion by KlatbahII** [moj-analytical-services/splink#1192] [General]
Hello, i tried to used this package by following the tutorial [Linking two tables of person](https://moj-analytical-services.github.io/splink/demos/example_link_only.html) and i have some issues.   

Here is the data i use
[pg_ak.csv](https://github.com/moj-analytical-services/splink/files/11257319/pg_ak.csv)
[shp_ak.csv](https://github.com/moj-analytical-services/splink/files/11257320/shp_ak.csv)

Here is the content of `pg_ak.csv`

| unique_id | hs_name        | hs_city   | hs_state |
| :-------- | :------------- | :-------- | :------- |
| 1         | Anchorage East | Anchorage | AK       |
| 2         | Juneau-Douglas | Juneau    | AK       |
| 3         | West Anchorage | Anchorage | AK       |
| 4         | Bartlett       | Anchorage | AK       |

And here the content of `shp_ak.csv`

| unique_id | hs_name                         | hs_city    | hs_state |
| :-------- | :------------------------------ | :--------- | :------- |
| 1         | Joann A. Alexie Memorial School | Atmautluak | AK       |
| 2         | Gladys Jung Elementary          | Bethel     | AK       |
| 3         | Bethel Regional High School     | Bethel     | AK       |
| 4         | Chaputnguak School              | Chefornak  | AK       |
| 5         | Chief Paul Memorial School      | Kipnuk     | AK       |
| 6         | Ayagina'ar Elitnaurvik          | Kongiganak | AK       |

There is around 500 lines in this CSV

What i want to do, i to get the matching entity of the dataset `shp_ak.csv` inside the dataset 
`pg_ak.csv`
There is one field in common between the two datasets : `hs_city`.  
This field is used as a blocking rule   
and i want to use the field `hs_name` to find the closest entity from the other dataset

Here is the code i used

```python
from splink.duckdb.duckdb_linker import DuckDBLinker
from splink.duckdb import duckdb_comparison_library as cl
import pandas as pd


df_qgis = pd.read_csv('shp_ak.csv')
df_pg = pd.read_csv('pg_ak.csv')

settings = {
    "link_type":"link_only",
    "blocking_rules_to_generate_predictions": [
        "l.hs_city = r.hs_city"
    ],
    "comparisons": [
        cl.levenshtein_at_thresholds("hs_name"),
        cl.exact_match("hs_city", term_frequency_adjustments=True)
    ]
}

linker = DuckDBLinker([df_pg, df_qgis], settings, input_table_aliases=["df_left","df_right"])
deterministic_rules =[
    "l.hs_city = r.hs_city and levenshtein(r.hs_name, l.hs_name) <= 5"
]

linker.estimate_probability_two_random_records_match(deterministic_rules, recall=0.7)
```

Here is the result of this code

```bash
WARNING: Deterministic matching rules led to no observed matches! This means that no possible record pairs are matches, and no records are linked to one another.
If this is truly the case then you do not need to run the linkage model.
However this is usually in error; expected rules to have recall of 70%. Consider revising rules as they may have an error.
Probability two random records match is estimated to be  0.
This means that amongst all possible pairwise record comparisons, one in Infinity are expected to match.  With 2,036 total possible comparisons, we expect a total of around 0.00 matching pairs
```

I wanted to know if i missed something, or did something wrong.  
Thanks a lot in advance

**Comment by RossKen** [comment]
Hey @KlatbahII, all this error is telling you is that there are no instances where "l.hs_city = r.hs_city and levenshtein(r.hs_name, l.hs_name) <= 5" is true. 

From a quick look, this appears to be because `df_qgis` includes names with "Memorial School", "Elementary", "High School", whereas `df_pg` does not. Levenshtein is the number of character additions, deletions or replacements to make two strings the same.  Given that "School", "Memorial School" and "Elementary" are all longer than your Levenshtein distance (5), no names between the two datasets will satisfy your condition.

I would recommend considering a more suitable string fuzzy match. Something like [jaro-winkler](https://moj-analytical-services.github.io/splink/topic_guides/comparators.html#jaro-winkler-similarity) which considers common characters between the strings, but places more weight towards the start of the string. In DuckDB the function name is `jaro_winkler_similarity`

For example:
```
from splink.duckdb.duckdb_linker import DuckDBLinker
from splink.duckdb import duckdb_comparison_library as cl
import pandas as pd


df_qgis = pd.read_csv('shp_ak.csv')
df_pg = pd.read_csv('pg_ak.csv')

settings = {
    "link_type":"link_only",
    "blocking_rules_to_generate_predictions": [],
    "comparisons": [
        cl.jaro_winkler_at_thresholds("hs_name", [0.8]),
        cl.exact_match("hs_city", term_frequency_adjustments=True)
    ]
}

linker = DuckDBLinker([df_pg, df_qgis], settings, input_table_aliases=["df_left","df_right"])
deterministic_rules =[
    "l.hs_city = r.hs_city and jaro_winkler_similarity(r.hs_name, l.hs_name) >= 0.8"
]

linker.estimate_probability_two_random_records_match(deterministic_rules, recall=0.7)
```

where I have also removed your blocking rules - given the small size of the data you can assess all pairwise comparisons

**Comment by KlatbahII** [comment]
Hi, thanks a lot for your response.  
I tried your code with bigger dataset and here is the message i get.  
So the problem i had is fixed

```bash
Probability two random records match is estimated to be  0.00112.
This means that amongst all possible pairwise record comparisons, one in 889.67 are expected to match.  
With 87,696 total possible comparisons, we expect a total of around 98.57 matching pairs
```

For the next steps, i follow the tutorials i found on documentation, and i just wanted to be sure i'm doing the right things

Then i do this step

```bash
linker.estimate_u_using_random_sampling(max_pairs=1e6)

----- Estimating u probabilities using random sampling -----
u probability not trained for hs_name - Exact match (comparison vector value: 2). This usually means the comparison level was never observed in the training data.

Estimated u probabilities using random sampling

Your model is not yet fully trained. Missing estimates for:
    - hs_name (some u values are not trained, no m values are trained).
    - hs_city (no m values are trained).
```

I continue by using the field in common between the two datasets

```bash
session_hs_city = linker.estimate_parameters_using_expectation_maximisation("l.hs_city = r.hs_city")

----- Starting EM training session -----

Estimating the m probabilities of the model by blocking on:
l.hs_city = r.hs_city

Parameter estimates will be made for the following comparison(s):
    - hs_name

Parameter estimates cannot be made for the following comparison(s) since they are used in the blocking rules: 
    - hs_city


WARNING:
Level Exact match on comparison hs_name not observed in dataset, unable to train m value
Iteration 1: Largest change in params was -0.95 in the m_probability of hs_name, level `Exact match`

WARNING:
Level Exact match on comparison hs_name not observed in dataset, unable to train m value
Iteration 2: Largest change in params was -0.0406 in the m_probability of hs_name, level `All other comparisons`

WARNING:
Level Exact match on comparison hs_name not observed in dataset, unable to train m value
Iteration 3: Largest change in params was 0.00234 in probability_two_random_records_match

WARNING:
Level Exact match on comparison hs_name not observed in dataset, unable to train m value
Iteration 4: Largest change in params was 0.000127 in probability_two_random_records_match

WARNING:
Level Exact match on comparison hs_name not observed in dataset, unable to train m value
Iteration 5: Largest change in params was 6.43e-06 in probability_two_random_records_match

EM converged after 5 iterations
m probability not trained for hs_name - Exact match (comparison vector value: 2). This usually means the comparison level was never observed in the training data.

Your model is not yet fully trained. Missing estimates for:
    - hs_name (some u values are not trained, some m values are not trained).
    - hs_city (no m values are trained).
```

Then i use the method you recommended

```bash
session_hs_name = linker.estimate_parameters_using_expectation_maximisation("jaro_winkler_similarity(r.hs_name, l.hs_name) >= 0.8")

----- Starting EM training session -----

Estimating the m probabilities of the model by blocking on:
jaro_winkler_similarity(r.hs_name, l.hs_name) >= 0.8

Parameter estimates will be made for the following comparison(s):
    - hs_city

Parameter estimates cannot be made for the following comparison(s) since they are used in the blocking rules: 
    - hs_name

Iteration 1: Largest change in params was -0.118 in probability_two_random_records_match
Iteration 2: Largest change in params was -0.00233 in probability_two_random_records_match
Iteration 3: Largest change in params was -0.000569 in the m_probability of hs_city, level `Exact match`
Iteration 4: Largest change in params was 0.000547 in the m_probability of hs_city, level `All other comparisons`
Iteration 5: Largest change in params was 0.000549 in the m_probability of hs_city, level `All other comparisons`
Iteration 6: Largest change in params was -0.000552 in the m_probability of hs_city, level `Exact match`
Iteration 7: Largest change in params was 0.000555 in the m_probability of hs_city, level `All other comparisons`
Iteration 8: Largest change in params was -0.000558 in the m_probability of hs_city, level `Exact match`
Iteration 9: Largest change in params was 0.000561 in the m_probability of hs_city, level `All other comparisons`
Iteration 10: Largest change in params was 0.000564 in the m_probability of hs_city, level `All other comparisons`
Iteration 11: Largest change in params was -0.000567 in the m_probability of hs_city, level `Exact match`
Iteration 12: Largest change in params was 0.000571 in the m_probability of hs_city, level `All other comparisons`
Iteration 13: Largest change in params was -0.000574 in the m_probability of hs_city, level `Exact match`
Iteration 14: Largest change in params was 0.000577 in the m_probability of hs_city, level `All other comparisons`
Iteration 15: Largest change in params was 0.00058 in the m_probability of hs_city, level `All other comparisons`
Iteration 16: Largest change in params was -0.000583 in the m_probability of hs_city, level `Exact match`
Iteration 17: Largest change in params was 0.000586 in the m_probability of hs_city, level `All other comparisons`
Iteration 18: Largest change in params was -0.000589 in the m_probability of hs_city, level `Exact match`
Iteration 19: Largest change in params was 0.000592 in the m_probability of hs_city, level `All other comparisons`
Iteration 20: Largest change in params was -0.000595 in the m_probability of hs_city, level `Exact match`
Iteration 21: Largest change in params was -0.000598 in the m_probability of hs_city, level `Exact match`
Iteration 22: Largest change in params was 0.000601 in the m_probability of hs_city, level `All other comparisons`
Iteration 23: Largest change in params was 0.000605 in the m_probability of hs_city, level `All other comparisons`
Iteration 24: Largest change in params was -0.000608 in the m_probability of hs_city, level `Exact match`
Iteration 25: Largest change in params was 0.000611 in the m_probability of hs_city, level `All other comparisons`

EM converged after 25 iterations

Your model is not yet fully trained. Missing estimates for:
    - hs_name (some u values are not trained, some m values are not trained).
```

And then i predict(i used a low threshold to see the results i could get)

```bash
results = linker.predict(threshold_match_probability=0.2)
results_df = results.as_pandas_dataframe()
```

And here is the results i get

|     | match_weight | match_probability | source_dataset_l | source_dataset_r | unique_id_l | unique_id_r | hs_name_l | hs_name_r                   | gamma_hs_name | hs_city_l | hs_city_r | gamma_hs_city |
| :-- | :----------- | :---------------- | :--------------- | :--------------- | :---------- | :---------- | :-------- | :-------------------------- | :------------ | :-------- | :-------- | :------------ |
| 0   | 7.619212     | 0.994940          | df_left          | df_right         | 1           | 761         | Abbeville | Abbeville High School       | 1             | Abbeville | Abbeville | 1             |
| 1   | 7.619212     | 0.994940          | df_left          | df_right         | 1           | 762         | Abbeville | Abbeville Elementary School | 1             | Abbeville | Abbeville | 1             |
| 2   | 5.619212     | 0.980061          | df_left          | df_right         | 3           | 120         | Athens    | Athens Intermediate School  | 1             | Athens    | Athens    | 1             |
| 3   | 5.619212     | 0.980061          | df_left          | df_right         | 3           | 121         | Athens    | Athens Elementary School    | 1             | Athens    | Athens    | 1             |
| 4   | 5.619212     | 0.980061          | df_left          | df_right         | 3           | 122         | Athens    | Athens High School          | 1             | Athens    | Athens    | 1             |
| 64  | 8.034250     | 0.996200          | df_left          | df_right         | 56          | 1359        | Vincent   | Vincent Elementary School   | 1             | Vincent   | Vincent   | 1             |
| 65  | 8.034250     | 0.996200          | df_left          | df_right         | 56          | 1360        | Vincent   | Vincent Middle High School  | 1             | Vincent   | Vincent   | 1             |
| 66  | 6.619212     | 0.989930          | df_left          | df_right         | 2           | 106         | Andalusia | Andalusia High School       | 1             | Andalusia | Andalusia | 1             |
| 67  | 6.619212     | 0.989930          | df_left          | df_right         | 2           | 107         | Andalusia | Andalusia Junior High       | 1             | Andalusia | Andalusia | 1             |
| 68  | 6.619212     | 0.989930          | df_left          | df_right         | 2           | 108         | Andalusia | Andalusia Elementary School | 1             | Andalusia | Andalusia | 1             |
|     |              |                   |                  |                  |             |             |           |                             |               |           |           |               |

The results are really great, thanks again.
I found it really good, but i didn't understand completly this part:

	Your model is not yet fully trained. Missing estimates for:
    - hs_name (some u values are not trained, some m values are not trained).

So i hope i could get your insights on the full process, and if i missed some steps.  
The library is awesome, thanks a lot again for your response

**Comment by RossKen** [comment]
Ah ok, I think I know why this is happening.

Again, this is coming from the fact that there are no instances where hs_name_l and hs_name_r are an exact match. From your warnings:
```
u probability not trained for hs_name - Exact match (comparison vector value: 2). This usually means the comparison level was never observed in the training data.
```
```
m probability not trained for hs_name - Exact match (comparison vector value: 2). This usually means the comparison level was never observed in the training data.
```
This has been introduced because `jaro_winkler_at_thresholds` in your comparison includes an exact match level. If you run `cl.jaro_winkler_at_thresholds("hs_name", 0.8).as_dict()` you can see the levels that come from it:
```
{'output_column_name': 'hs_name',
 'comparison_levels': [
  {'sql_condition': '"hs_name_l" IS NULL OR "hs_name_r" IS NULL',
   'label_for_charts': 'Null',
   'is_null_level': True},
  {'sql_condition': '"hs_name_l" = "hs_name_r"',
   'label_for_charts': 'Exact match'},
  {'sql_condition': 'jaro_winkler_similarity("hs_name_l", "hs_name_r") >= 0.8',
   'label_for_charts': 'Jaro_winkler_similarity >= 0.8'},
  {'sql_condition': 'ELSE', 'label_for_charts': 'All other comparisons'}],
 'comparison_description': 'Exact match vs. jaro_winkler_similarity at threshold 0.8 vs. anything else'}
```
There is a parameter to remove this exact match level which should solve your issues:
```
jaro_winkler_at_thresholds("hs_name", 0.8, include_exact_match_level=False)
```
I hope this helps!
